x-pack/plugins/fleet/server/services/epm/packages/cleanup.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { SavedObjectsClientContract } from 'src/core/server';
+
+import { removeArchiveEntries } from '../archive/storage';
+
+import { ASSETS_SAVED_OBJECT_TYPE, PACKAGE_POLICY_SAVED_OBJECT_TYPE } from '../../../../common';
+import type { PackageAssetReference } from '../../../../common';
+import { packagePolicyService } from '../../package_policy';
+import { appContextService } from '../..';
+
+export async function removeOldAssets(options: {
+  savedObjectsClient: SavedObjectsClientContract;
+  pkgName: string;
+  oldVersion: string;
```

Comment:
I think we may want to change this API to handle the scenario where there are multiple outdated versions that haven't been cleaned up. This would help cover the case where customers have been using Fleet for a while before they upgrade to the version that has this cleanup logic. So instead, I'd expect that we pass in the `currentVersion` and then do a best effort to cleanup all older assets that are not in use by a package policy.

Reply:
yes, I was planning to extend as a next step.
do you think we can run the same cleanup logic on package update/policy upgrade for all older versions? so we don't need two versions of cleanup.
regarding the implementation, I was thinking to use the find api to find all lower version assets and check for each version that there are no dependent policies. would that work with an aggregation to find distinct versions?

Reply:
I updated this to find all older versions and delete assets for each. Tested with multiple old versions, seems to be working fine.

Reply:
> do you think we can run the same cleanup logic on package update/policy upgrade for all older versions? so we don't need two versions of cleanup.

I can't think of a reason it wouldn't work for both scenarios. It would be good to reuse the same logic for each.

> regarding the implementation, I was thinking to use the find api to find all lower version assets and check for each version that there are no dependent policies. would that work with an aggregation to find distinct versions?

An aggregation makes total sense to me. The find API does support aggregations as well, let me know if you have any trouble, but I suspect a terms aggregation on the `package_policy` field should yield what you're trying to achieve. This will give you the number of documents that match each unique version keyword.

Reply:
yes, I found that, it is working